src/main/java/org/candlepin/model/AbstractHibernateCurator.java

```diff
@@ -255,17 +255,44 @@ public E create(E entity, boolean flush) {
     /**
      * @return all entities for a particular type.
      */
-    public CandlepinQuery<E> listAll() {
-        DetachedCriteria criteria = this.createSecureDetachedCriteria();
+    public List<E> listAll() {
+        EntityManager em = this.getEntityManager();
+        CriteriaBuilder criteriaBuilder = em.getCriteriaBuilder();
+        CriteriaQuery<E> criteriaQuery = criteriaBuilder.createQuery(this.entityType);
+        Root<E> root = criteriaQuery.from(this.entityType);
+        criteriaQuery.select(root);
+
+        List<Predicate> predicates = new ArrayList<>();
+        Predicate securityPredicate = this.getSecurityPredicate(this.entityType, criteriaBuilder, root);
+
+        if (securityPredicate != null) {
+            predicates.add(securityPredicate);
+            criteriaQuery.where(predicates.toArray(new Predicate[predicates.size()]));
+        }
 
-        return this.cpQueryFactory.buildQuery(this.currentSession(), criteria);
+        return em.createQuery(criteriaQuery).getResultList();
     }
 
-    public CandlepinQuery<E> listAllByIds(Collection<? extends Serializable> ids) {
-        DetachedCriteria criteria = this.createSecureDetachedCriteria()
-            .add(CPRestrictions.in("id", ids));
+    public List<E> listAllByIds(Collection<? extends Serializable> ids) {
+        EntityManager em = this.getEntityManager();
+        CriteriaBuilder criteriaBuilder = em.getCriteriaBuilder();
+        CriteriaQuery<E> criteriaQuery = criteriaBuilder.createQuery(this.entityType);
+        Root<E> root = criteriaQuery.from(this.entityType);
+        criteriaQuery.select(root);
+
+        CriteriaBuilder.In<String> inClause = criteriaBuilder.in(root.get("id"));
+        ids.stream().forEach(id -> inClause.value(id.toString()));
```

Comment:
I think you can do something like `Predicate idsInPredicate = inClause.in(ids);` instead of looping over all the ids and then you can add `idsInPredicate` to the `predicates` list. There is probably a better name than idsInPredicate, I was just giving an example on this!

Reply:
To add to this, it's not strictly necessary to maintain a list of predicates, since we have a known predicate with the IN clause, and we *may* have one with the "security" predicate. So we can avoid the list with a single predicate as follows:

```
Predicate predicate = ...

Predicate securityPredicate = ...
if (securityPredicate != null) {
  predicate = criteriaBuilder.and(predicate, securityPredicate);
}

criteriaQuery.where(predicate);
```